File: src/python/yeti.py
from browser import window

from enemy import Enemy
from player import Player
import ika
import Brain
import sound

_anim = {
    'idle': ((
        ((7, 1000),),
        ((14, 1000),),
        ((21, 1000),),
        ((0, 1000),),
        ((7, 1000),),
        ((14, 1000),),
        ((7, 1000),),
        ((14, 1000),),
        ),
        True
    ),

    'walk': ((
        window.animator.makeAnimRange(8, 14, 15),
        window.animator.makeAnimRange(15, 21, 15),
        window.animator.makeAnimRange(22, 28, 15),
        window.animator.makeAnimRange(1, 7, 15),
        window.animator.makeAnimRange(8, 14, 15),
        window.animator.makeAnimRange(15, 21, 15),
        window.animator.makeAnimRange(8, 14, 15),
        window.animator.makeAnimRange(15, 21, 15),
        ),
        True
    ),

    'hurt': ((
        ((45, 1000),),
        ((38, 1000),),
        ((31, 1000),),
        ((52, 1000),),
        ((45, 1000),),
        ((38, 1000),),
        ((45, 1000),),
        ((38, 1000),),
        ),
        True
    ),

    'die': ((
        ((38, 30), (39, 90)),
        ((45, 30), (46, 90)),
        ((52, 30), (53, 90)),
        ((31, 30), (32, 90)),
        ((38, 30), (39, 90)),
        ((45, 30), (46, 90)),
        ((38, 30), (39, 90)),
        ((45, 30), (46, 90)),
        ),
        False
    ),

    'windup': ((
        ((35, 25),(36, 10)),
        ((42, 25),(43, 10)),
        ((49, 25),(50, 10)),
        ((28, 25),(29, 10)),
        ((35, 25),(36, 10)),
        ((42, 25),(43, 10)),
        ((35, 25),(36, 10)),
        ((42, 25),(43, 10)),
        ),
        False
    ),

    'attack': ((
        ((37, 20),),
        ((44, 20),),
        ((51, 20),),
        ((30, 20),),
        ((37, 20),),
        ((44, 20),),
        ((37, 20),),
        ((44, 20),),
        ),
        False
    ),
}

_attackRange = (
    (-24, 0, 24, 32),
    (32, 0, 24, 32),
    (-22, -24, 52, 24),
    (-22, 32, 52, 24),
    (-24, 0, 24, 32),
    (32, 0, 24, 32),
    (-24, 0, 24, 32),
    (32, 0, 24, 32),
)

class Yeti(Enemy):
    def __init__(self, engineRef, ent):
        Enemy.__init__(self, engineRef, ent, _anim, Brain.Brain())
        self.speed = 80

        self.addMoods(
            (Brain.Attack(1), self.attackMood),
            (Brain.Passive(1), self.passiveMood)
        )

        self.stats.maxhp = 100
        self.stats.hp = self.stats.maxhp
        self.stats.att = 20
        self.stats.exp = 100
        self.stats.ind = 0

    def hurtState(self, dist, dir):
        if self.stats.hp > 0:
            i = ika.Random(0, len(sound.yetiHurt[self.stats.ind]))
            sound.yetiHurt[self.stats.ind][i].Play()

        self.mood = self.attackMood
        yield from super(Yeti, self).hurtState(dist * 2 // 3, dir)

    def attackMood(self):
        # if we want to be uber, we can remove this hack.
        # for now fuckit.  Attack the player!!
        p = self.engineRef.player
        for q in range(5):
            # compensate for the yeti's gigantic sprite:
            sx = self.x + 16
            sy = self.y + 16
            d = self.engineRef.dir.fromDelta(p.x - sx, p.y - sy)
            dist = ika.hypot(p.x - sx, p.y - sy)
            if dist < 50:
                pass
                yield self.attackState(d)
                yield self.idleState(20)
            else:
                yield self.walkState(d, min(90, dist))

    def passiveMood(self):
        p = self.engineRef.player
        self.stopAnimation()
        while True:
            # compensate for the yeti's gigantic sprite:
            sx = self.x + 16
            sy = self.y + 16
            dist = ika.hypot(p.x - sx, p.y - sy)

            yield self.idleState()

            if dist < 150:
                self.mood = self.attackMood
                yield self.idleState()
                break
            brython_generator_bug_workaround = 'blah'

    def deathState(self, *args, **kwargs):
        sound.yetiDie[self.stats.ind].Play()
        self.startAnimation('die')
        yield from super(Yeti, self).deathState(*args, **kwargs)

    def walkState(self, dir, dist):
        self.move(dir, dist)
        self.startAnimation('walk')
        dist *= 100
        while dist > 0:
            dist -= self.speed
            yield None
        self.stop()

    def attackState(self, dir):
        oldInterruptable = self.interruptable
        def restoreVars(self=self, oldInterruptable=oldInterruptable):
            self.interruptable = oldInterruptable
        self._onStateExit = restoreVars

        self.direction = dir
        self.startAnimation('windup')
        self.stop()
        sound.yetiStrike[self.stats.ind].Play()
        self.interruptable = False

        # Wind up.  Hold up a sec.
        for i in range(35):
            yield None

        self.startAnimation('attack')
        self.move(dir, 6)
        for i in range(20):
            for e in self.detectCollision(_attackRange[dir]):
                if isinstance(e, Player):
                    d = max(1, self.stats.att - e.stats.pres)
                    e.hurt(d, 350, self.direction)
            yield None

        self.stop()

        self.state = self.idleState(10)
        yield None

File: src/winter/yeti.ts
import { makeAnimRange } from "./animator.js"
import { Direction, fromDelta } from "./Direction.js"
import { Enemy } from "./enemy.js"
import { hypot, PyEngine, Sprite } from "./winter.js"

const _anim: {[key: string]: [[number, number][][], boolean]} = {
    idle: [[
        [[7, 1000]],
        [[14, 1000]],
        [[21, 1000]],
        [[0, 1000]],
        [[7, 1000]],
        [[14, 1000]],
        [[7, 1000]],
        [[14, 1000]],
    ], true],

    walk: [[
        makeAnimRange(8, 14, 15),
        makeAnimRange(15, 21, 15),
        makeAnimRange(22, 28, 15),
        makeAnimRange(1, 7, 15),
        makeAnimRange(8, 14, 15),
        makeAnimRange(15, 21, 15),
        makeAnimRange(8, 14, 15),
        makeAnimRange(15, 21, 15),
    ], true],

    hurt: [[
        [[45, 1000]],
        [[38, 1000]],
        [[31, 1000]],
        [[52, 1000]],
        [[45, 1000]],
        [[38, 1000]],
        [[45, 1000]],
        [[38, 1000]],
    ], true],

    die: [[
        [[38, 30], [39, 90]],
        [[45, 30], [46, 90]],
        [[52, 30], [53, 90]],
        [[31, 30], [32, 90]],
        [[38, 30], [39, 90]],
        [[45, 30], [46, 90]],
        [[38, 30], [39, 90]],
        [[45, 30], [46, 90]],
    ], false],

    windup: [[
        [[35, 25],[36, 10]],
        [[42, 25],[43, 10]],
        [[49, 25],[50, 10]],
        [[28, 25],[29, 10]],
        [[35, 25],[36, 10]],
        [[42, 25],[43, 10]],
        [[35, 25],[36, 10]],
        [[42, 25],[43, 10]],
    ], false],

    attack: [[
        [[37, 20],],
        [[44, 20],],
        [[51, 20],],
        [[30, 20],],
        [[37, 20],],
        [[44, 20],],
        [[37, 20],],
        [[44, 20],],
    ], false],
}

const _attackRange: [number, number, number, number][] = [
    [-24, 0, 24, 32],
    [32, 0, 24, 32],
    [-22, -24, 52, 24],
    [-22, 32, 52, 24],
    [-24, 0, 24, 32],
    [32, 0, 24, 32],
    [-24, 0, 24, 32],
    [32, 0, 24, 32],
]

export class Yeti extends Enemy {
    protected dieSound: string
    protected hurtSound: string
    protected strikeSound: string

    constructor(
        engineRef: PyEngine,
        sprite: Sprite,
    ) {
        super(engineRef, sprite, _anim)
        this.sprite.speed = 80

        this.addMoods([this.attackMood.bind(this), this.passiveMood.bind(this)])

        this.stats.maxhp = 100
        this.stats.hp = this.stats.maxhp
        this.stats.att = 20
        this.stats.exp = 100
        this.dieSound = 'yetiDie'
        this.hurtSound = 'yetiHurt'
        this.strikeSound = 'yetiStrike'
    }

    *hurtState(dist: number, dir: Direction) {
        if (this.stats.hp > 0) {
            this.engine.sounds[this.hurtSound].play()
        }

        this.setMood(this.attackMood.bind(this))
        yield* super.hurtState(Math.floor(dist * 2 / 3), dir)
    }

    *attackMood() {
        // if we want to be uber, we can remove this hack.
        // for now fuckit.  Attack the player!!
        const p = this.engineRef.getPlayerEntity()
        for (let q = 0; q < 5; ++q) {
            // compensate for the yeti's gigantic sprite:
            const sx = this.sprite.x + 16
            const sy = this.sprite.y + 16
            const d = fromDelta(p.sprite.x - sx, p.sprite.y - sy)
            const dist = hypot(p.sprite.x - sx, p.sprite.y - sy)
            if (dist < 50) {
                yield this.attackState(d)
                yield this.idleState(20)
            } else {
                yield this.walkState(d, Math.min(90, dist))
            }
        }
    }

    *passiveMood() {
        const p = this.engineRef.getPlayerEntity()
        this.stopAnimation()
        while (true) {
            // compensate for the yeti's gigantic sprite:
            const sx = this.sprite.x + 16
            const sy = this.sprite.y + 16
            const dist = hypot(p.sprite.x - sx, p.sprite.y - sy)

            yield this.idleState()

            if (dist < 150) {
                this.setMood(this.attackMood.bind(this))
                yield this.idleState()
                return
            }
        }
    }

    *deathState() {
        this.engine.sounds[this.dieSound].play()
        this.startAnimation('die')
        yield* super.deathState()
    }

    *walkState(dir: Direction, dist: number) {
        this.move(dir, dist)
        this.startAnimation('walk')
        dist *= 100
        while (dist > 0) {
            dist -= this.sprite.speed
            yield null
        }
        this.stop()
    }

    *attackState(dir: Direction) {
        const oldInterruptable = this.interruptable
        this._onStateExit = () => {
            this.interruptable = oldInterruptable
        }

        this.direction = dir
        this.startAnimation('windup')
        this.stop()
        this.engine.sounds[this.strikeSound].play()
        this.interruptable = false

        // Wind up.  Hold up a sec.
        for (let i = 0; i < 35; ++i) {
            yield null
        }

        this.startAnimation('attack')
        this.move(dir, 6)
        for (let i = 0; i < 20; ++i) {
            for (let e of this.detectCollision(_attackRange[dir])) {
                if (e.isKind('Player')) {
                    const d = Math.max(1, this.stats.att - e.stats.pres)
                    e.hurt(d, 350, this.direction)
                }
            }
            yield null
        }

        this.stop()

        this.setState(this.idleState(10))
        yield null
    }
}

export class Gorilla extends Yeti {
    constructor(
        engineRef: PyEngine,
        sprite: Sprite,
    ) {
        super(engineRef, sprite)
        this.sprite.speed = 90
        this.stats.maxhp = 300
        this.stats.hp = 200
        this.stats.att = 36
        this.stats.exp = 200
    }
}

export class SoulReaver extends Yeti {
    constructor(
        engineRef: PyEngine,
        sprite: Sprite,
    ) {
        super(engineRef, sprite)
        this.sprite.speed = 120
        this.stats.maxhp = 1500
        this.stats.hp = 1500
        this.stats.att = 60
        this.stats.exp = 750
        this.dieSound = 'soulReaverDie'
        this.hurtSound = 'soulReaverHurt'
        this.strikeSound = 'soulReaverStrike'
    }
        
    *attackState(direc: Direction) {
        this.direction = direc
        if (this.direction === Direction.UpLeft || this.direction == Direction.DownLeft) {
            this.direction = Direction.Left
        } else if (this.direction == Direction.UpRight || this.direction == Direction.DownRight) {
            this.direction = Direction.Right
        }

        const oldInterruptable = this.interruptable
        this._onStateExit = () => {
            this.interruptable = oldInterruptable
        }

        this.startAnimation('windup')
        this.stop()
        this.engine.sounds[this.strikeSound].play()
        this.interruptable = false

        // Wind up.  Hold up a sec.
        // Show first frame for a bit longer than usual.
        this.stopAnimation()
        for (let i = 0; i < 75; ++i) {
            yield null
        }
        this.startAnimation('windup')
        for (let i = 0; i < 35; ++i) {
            yield null
        }

        this.startAnimation('attack')
        this.sprite.speed += 800
        this.move(this.direction, 2000)

        for (let i = 0; i < 8; ++i) {
            this.sprite.speed -= (i + 2) * 10
            for (let e of this.detectCollision(_attackRange[this.direction])) {
                if (e.isKind('Player')) {
                    const d = Math.max(1, this.stats.att - e.stats.pres)
                    e.hurt(d, 350, this.direction)
                }
            }
            yield null
        }

        for (let i = 0; i < 30; ++i) {
            this.sprite.speed = Math.max(10, this.sprite.speed - 10)
            yield null
        }

        this.stop()

        this.setState(this.idleState(10))
        this.sprite.speed = 120

        yield null
    }
}
